src/Gateway/Utils/Middleware/ErrorHandler.cs

```diff
@@ -0,0 +1,72 @@
+﻿using Gateway.Models.Exceptions;
+using Microsoft.AspNetCore.Builder;
+using Microsoft.AspNetCore.Http;
+using Newtonsoft.Json;
+using System;
+using System.Net;
+using System.Threading.Tasks;
+
+namespace Gateway.Utils.Middleware
+{
+    // You may need to install the Microsoft.AspNetCore.Http.Abstractions package into your project
+    public class ErrorHandler
+    {
+        private readonly RequestDelegate _next;
+
+        public ErrorHandler(RequestDelegate next)
+        {
+            _next = next;
+        }
+
+        public async Task Invoke(HttpContext httpContext)
+        {
+            try
+            {
+                await _next(httpContext);
+            }
+            catch(Exception ex)
+            {
+                await HandleExceptionAsync(httpContext, ex);
+            }
+        }
+
+        /// <summary>
+        /// Handles exceptions, outputting an anonymous object
+        /// </summary>
+        /// <param name="context">Http context containing a response</param>
+        /// <param name="exception">The exception to return to the user</param>
+        /// <returns>502 error (Bad Gateway) with full error enclosed within the body</returns>
+        private async Task HandleExceptionAsync(HttpContext context, Exception exception)
+        {
+            var response = context.Response;
+            var statusCode = (int)HttpStatusCode.BadGateway;
+            var message = "Unexpected error";
+            var description = "Unexpected Error";            
+
+            if(exception is CustomException customException)
+            {
+                message = customException.Message;
+                description = customException.Description;
+                statusCode = customException.StatusCode;
+            }
+
+            response.ContentType = "application/json";
+            response.StatusCode = (int)HttpStatusCode.BadGateway;
+            await response.WriteAsync(JsonConvert.SerializeObject(new {
+                Message = message,
+                Description = description,
+                StatusCode = statusCode,
+                exception.StackTrace
+            }));
```

Comment:
I don't think it would be a good idea to return BadGateway status for a request downstream that returned a 404 not found. Some HTTP clients rely on the status code and potentially will ignore the body.

Could we mimic the status from the downstream request?

Reply:
Sorted in latest commits. Now passes on resultant status code from exception